### PR TITLE
Add Boolean Logic

### PR DESCRIPTION
**Pull request by hughjazzman**
* Update `BooleanParser`
* Update `BooleanCommand` subclasses

For @sevenseasofbri to work on `CalculateBooleanCommand`.

### CODE REVIEW

File: src/main/java/seedu/duke/model/binarytree/BinaryTree.java [hunk 1]
```diff
@@ -34,6 +34,16 @@ public T getRoot() {
...
     }
 
+    /**
+     * Returns T object at index of tree.
+     *
+     * @param idx Index in tree.
+     * @return T object.
+     */
+    public T getT(int idx) {
+        return arrayList.get(idx);
```
Comment (sevenseasofbri) [inline src/main/java/seedu/duke/model/binarytree/BinaryTree.java:48]: We could assert that the logic will not call anything out of bounds?

**Review by sevenseasofbri — APPROVED** [review]
Just a minor suggestion (that too for testing/defensiveness of code) otherwise LGTM!!

File: src/main/java/seedu/duke/model/binarytree/BinaryTree.java (hunk 1 above)
Comment (hughjazzman) [reply src/main/java/seedu/duke/model/binarytree/BinaryTree.java:48]: Good point. Should there be an assertion in `BooleanTemplate.setInput()` as well? Though it does seem we are using `DukeException` for `BinaryTree.insert()`. Maybe all as exceptions?

Comment (sevenseasofbri) [reply src/main/java/seedu/duke/model/binarytree/BinaryTree.java:48]: Yup, sounds good!

**Pull request merged**